File: src/day4.rs
use crate::util::read_file;
use fancy_regex::Regex;

pub fn run() {
    println!("---- DAY 4 ----");
    part_1();
    part_2();
}

fn reg(code: &str) -> Regex {
    Regex::new(format!("{}:(?<data>[^ ]*)", code).as_str()).unwrap()
}

fn part_1() {
    let lines = read_file("inputs/day4.txt");
    let fields = vec!["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"];
    let mut valid = 0;
    let mut found: Vec<&str> = vec![];
    let mut passport = String::new();
    for line in lines.split('\n') {
        if line == "" {
            // validate passport
            for field in &fields {
                let r = reg(field);
                if r.is_match(&passport).unwrap() {
                    found.push(field);
                }
            }
            found.dedup();
            if found.len() == fields.len() {
                valid += 1;
            }
            found = vec![];
            passport.clear();
            continue
        }
        passport.push_str(format!(" {}", line).as_str());
    }
    println!("There are [{}] valid passports", valid);
}

fn is_valid_data(field: &str, reg: Regex, passport: &str) -> bool {
    let groups = reg.captures(passport).unwrap().unwrap();
    let data = groups.name(&"data").unwrap().as_str();
    match field {
        "byr" => {
            let i = data.parse::<i32>().unwrap();
            i >= 1920 && i <= 2002
        },
        "iyr" => {
            let i = data.parse::<i32>().unwrap();
            i >= 2010 && i <= 2020
        },
        "eyr" => {
            let i = data.parse::<i32>().unwrap();
            i >= 2020 && i <= 2030
        },
        "hgt" => {
            let result = Regex::new(r"^(?<i>\d+)(?<m>cm|in)$")
                .unwrap()
                .captures(data)
                .unwrap();
            match result {
                Some(groups) => {
                    let i: i32 = groups.name(&"i").unwrap().as_str().parse().unwrap();
                    let m = groups.name(&"m").unwrap().as_str();
                    if m == "cm" {
                        i >= 150 && i <= 193
                    } else if m == "in" {
                        i >= 59 && i <= 76
                    } else {
                        false
                    }
                },
                _ => false
            }
        },
        "hcl" => Regex::new(r"^#[0-9a-f]{6}$").unwrap().is_match(data).unwrap(),
        "ecl" => Regex::new(r"^(amb|blu|brn|gry|grn|hzl|oth)$").unwrap().is_match(data).unwrap(),
        "pid" => Regex::new(r"^\d{9}$").unwrap().is_match(data).unwrap(),
        _ => true
    }
}

fn part_2() {
    let lines = read_file("inputs/day4.txt");
    let fields = vec!["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"];
    let mut valid = 0;
    let mut found: Vec<&str> = vec![];
    let mut passport = String::new();
    for line in lines.split('\n') {
        if line == "" {
            // validate passport
            for field in &fields {
                let r = reg(field);
                if r.is_match(passport.as_str()).unwrap() {
                    if is_valid_data(*field, r, passport.as_str()) {
                        found.push(field);
                    }
                }
            }
            found.dedup();
            if found.len() == fields.len() {
                valid += 1;
            }
            found.clear();
            passport.clear();
            continue;
        }
        passport.push_str(format!(" {}", line).as_str());
    }

    println!("There are [{}] passports with valid data", valid);
}

File: elixir/lib/day4.ex
defmodule Day4 do
    def run do
        IO.puts "---- DAY 4 ----"
        lines = Input.readFile("day4") 
            |> Enum.chunk_by(&(&1 == ""))
            |> Enum.filter(&(&1 != [""]))
            |> Enum.map(&(Enum.join(&1, " ")))
        fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
            |> Enum.reduce(%{}, fn(f, m) -> 
                case Regex.compile("#{f}:(?<data>[^ ]*)") do
                    {:ok, r} -> Map.put(m, f, r)
                    _ -> :nil
                end
            end)
        day1(lines, fields)
        day2(lines, fields)
    end

    defp day1(lines, fields) do
        validPassports = lines
            |> Enum.reduce(0, fn(line, valid) ->
                entries = line
                    |> String.split
                    |> Enum.reduce(0, fn(entry, found) ->
                        case fields |> Map.values |> Enum.filter(&(Regex.match?(&1, entry))) |> Enum.count do
                            0 -> found
                            _ -> found + 1
                        end
                    end)
                if entries < Enum.count(fields) do valid else valid + 1 end
            end)
        IO.puts "There are #{validPassports} valid passports"
    end

    defp inRange?(data, range) do String.to_integer(data) in range end

    defp fieldValid?(field, reg, entry) do
        data = reg |> Regex.named_captures(entry) |> Map.get("data")
        case field do
            "byr" -> inRange?(data, 1920..2002)
            "iyr" -> inRange?(data, 2010..2020)
            "eyr" -> inRange?(data, 2020..2030)
            "hgt" -> 
                case Regex.named_captures(~r/^(?<i>\d+)(?<m>cm|in)$/, data) do
                    :nil -> false
                    caps ->
                        case Map.get(caps, "m") do
                            "cm" -> inRange?(Map.get(caps, "i"), 150..193)
                            "in" -> inRange?(Map.get(caps, "i"), 59..76)
                            _ -> false
                        end
                end
            "hcl" -> Regex.match?(~r/^#[0-9a-f]{6}$/, data)
            "ecl" -> Regex.match?(~r/^(amb|blu|brn|gry|grn|hzl|oth)$/, data)
            "pid" -> Regex.match?(~r/^\d{9}$/, data)
        end
    end

    defp day2(lines, fields) do
        validPassports = lines
            |> Enum.filter(fn(line) ->
                line
                    |> String.split
                    |> Enum.filter(fn(entry) ->
                        fields
                            |> Map.keys
                            |> Enum.reduce(false, fn(f, found) ->
                                found or (Regex.match?(Map.get(fields, f), entry) and fieldValid?(f, Map.get(fields, f), entry))
                            end)
                    end)
                    |> Enum.count == fields |> Map.keys |> Enum.count
            end)
            |> Enum.count
        IO.puts "There are #{validPassports} ACTUALLY valid passports"
    end
end